src/webnn/native/dml/ContextDML.cpp

```diff
@@ -0,0 +1,74 @@
+// Copyright 2022 The WebNN-native Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "webnn/native/dml/ContextDML.h"
+
+#include "webnn/native/dml/GraphDML.h"
+
+namespace webnn::native::dml {
+
+    HRESULT Context::Initialize() {
+        if (mUseDebugLayer) {
+            ComPtr<ID3D12Debug> debug;
+            if (SUCCEEDED(D3D12GetDebugInterface(IID_PPV_ARGS(&debug)))) {
+                debug->EnableDebugLayer();
+            }
+        }
+        WEBNN_RETURN_IF_FAILED(D3D12CreateDevice(mAdapter.Get(), D3D_FEATURE_LEVEL_11_0,
+                                                 IID_PPV_ARGS(&mCommandRecorderDML.D3D12Device)));
+        D3D12_COMMAND_QUEUE_DESC commandQueueDesc{};
+        commandQueueDesc.Type = D3D12_COMMAND_LIST_TYPE_DIRECT;
+        commandQueueDesc.Flags = D3D12_COMMAND_QUEUE_FLAG_NONE;
+        WEBNN_RETURN_IF_FAILED(mCommandRecorderDML.D3D12Device->CreateCommandQueue(
```

Comment:
We can't have the context (itself) always create the [DML, D3D12] device and queue. Rather we want the context to initialize and execute with any (device, queue) specified.

Reply:
I don't understand, could you explain the purpose? @bbernhar 
These D3D related members will only be created once when we create one context, we don't always create them. And I think one context owns one CommandRecorder I defined  and the context will create multiple graphs, then graphs will do InitializeGraph and ExecuteGraph. Am I right? @fujunwei

Reply:
The [DML, D3D12] device and queue does not have a 1:1 relationship with a context. You can have multiple contexts per (device, queue). Chromium could, for example, specify a (device, queue) for this context. 

Should the context create it's own (device, queue, then we'll have to deal with cross-queue usage rules because it wouldn't necessarily be compatible.

A CommandRecorderDML and CommandRecorder has 1:1 relationship, the context does not care about recording, just execution of the underlying command-list the recorder produces.

Hopefully that makes more sense.

Reply:
Make sense, we should enum all adapters (integrateGPU, discreteGPU, maybe VPU) at once in the GPU process for all contexts like WebGPU's [Backend](https://source.chromium.org/chromium/chromium/src/+/main:third_party/dawn/src/dawn/native/d3d12/BackendD3D12.h;drc=8f90d2b35c2ae2bf6376493c6fd6aedcd0e2d0d1;l=29)::[DiscoverAdapters](https://source.chromium.org/chromium/chromium/src/+/main:third_party/dawn/src/dawn/native/d3d12/BackendD3D12.cpp;bpv=1;bpt=1;l=174?gsn=DiscoverAdapters&gs=kythe%3A%2F%2Fchromium.googlesource.com%2Fchromium%2Fsrc%3Flang%3Dc%252B%252B%3Fpath%3Dthird_party%2Fdawn%2Fsrc%2Fdawn%2Fnative%2Fd3d12%2FBackendD3D12.cpp%23fZROI_IpUZrumbshEGduqYYsEXl1mRvxB10w-d3iO3A&gs=kythe%3A%2F%2Fchromium.googlesource.com%2Fchromium%2Fsrc%3Flang%3Dc%252B%252B%3Fpath%3Dthird_party%2Fdawn%2Fsrc%2Fdawn%2Fnative%2Fd3d12%2FBackendD3D12.h%239hLYNblPZ5gc7d3uaRGbDrabZKm6k2KfJJo6AV5CT6c).

For each adapter in DML backend, there is a device(DML, D3D12) as member variable like WebGPU's [mD3d12Device](https://source.chromium.org/chromium/chromium/src/+/main:third_party/dawn/src/dawn/native/d3d12/AdapterD3D12.h;bpv=1;bpt=1;l=56?gsn=mD3d12Device&gs=kythe%3A%2F%2Fchromium.googlesource.com%2Fchromium%2Fsrc%3Flang%3Dc%252B%252B%3Fpath%3Dthird_party%2Fdawn%2Fsrc%2Fdawn%2Fnative%2Fd3d12%2FAdapterD3D12.h%23_brXG6ch6SRR3iRz1pkf2NSVey604QFZ6pcjZ15ppXw).

The device in DML backend has 1 queue and 1 commandRecorder to recoder commands for all contexts, the commands come from initialize and execute Graph, am i right?

Reply:
@fujunwei Only the queue is sharable. Since no other recorder can be provided to this context, it doesn't need to outlive it.

Reply:
Make sense, thanks for explain patiently.